sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -434,153 +418,129 @@ class GpuDynamicPartitionDataSingleWriter(
-          val previousPartPath = currentPartPath
-          currentPartPath = partPath
-
-          // see a new partition, close the old writer
-          val previousWriterStatus = currentWriterStatus
-          if (previousWriterStatus != null) {
-            releaseWriter(previousWriterStatus.outputWriter)
-          }
-
-          if (cachesMap.isDefined) {
-            savedStatus = cachesMap.get.get(currentPartPath)
-            if (savedStatus.isDefined) {
-              // first try to restore the saved writer status,
-              // `GpuDynamicPartitionDataConcurrentWriter` may already opened the writer, and may
-              // have pending caches
-              currentWriterStatus = savedStatus.get.writerStatus
-              // entire batch that is sorted, see a new partition, the old write status is useless
-              cachesMap.get.remove(previousPartPath)
-            } else {
-              // create a new one
-              val writer = newWriter(partPath, None, 0)
-              currentWriterStatus = new WriterStatus(writer)
-              statsTrackers.foreach(_.newPartition())
+    // We have an entire batch that is sorted, so we need to split it up by key
+    val (paths, splits) = withResource(cb) { _ =>
+      withResource(getPartitionColumnsAsTable(cb)) { partitionColumnsTbl =>
+        withResource(distinctAndSort(partitionColumnsTbl)) { distinctKeysTbl =>
+          val partitionIndexes = splitIndexes(partitionColumnsTbl, distinctKeysTbl)
+          withResource(copyToHostAsBatch(distinctKeysTbl, partDataTypes)) { cbKeys =>
+            val splits = withResource(getOutputColumnsAsTable(cb)) { outputColumnsTbl =>
+              outputColumnsTbl.contiguousSplit(partitionIndexes: _*)
+            }
+            closeOnExcept(splits) { _ =>
+              // Use the existing code to convert each row into a path. It would be nice to do this
+              // on the GPU, but the data should be small and there are things we cannot easily
+              // support on the GPU right now
+              import scala.collection.JavaConverters._
+              // paths
+              val paths = cbKeys.rowIterator().asScala.map(getPartitionPath).toArray
+              (paths, splits)
             }
-          } else {
-            // create a new one
-            val writer = newWriter(partPath, None, 0)
-            currentWriterStatus = new WriterStatus(writer)
-            statsTrackers.foreach(_.newPartition())
           }
         }
+      }
+    }
+    withResource(splits) { _ =>
+      paths.zipWithIndex.foreach { case (partPath, partIx) =>
+        val partContigTable = splits(partIx)
+        // If fall back from for `GpuDynamicPartitionDataConcurrentWriter`, we should get the
+        // saved status
+        val savedStatus = updateCurrentWriterIfNeeded(partPath, cachesMap)
 
         if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
           // convert caches seq to tables and close caches seq
```

Comment:
nit: Could we split this large function up into smaller functions?  It is quite hard to see what the else part of this if goes with.

There also appears to be a lot of common code between the various paths and it would be nice to make it common, or at least see what the differences are better. Like updating the currentWriters status and metrics/etc.

Reply:
sure thing